internal/pkg/consul/client.go

```diff
@@ -0,0 +1,347 @@
+// Sends the configuration in the target struct as interface{} on updateChannel, which caller must cast
+func (client *consulClient) WatchForChanges(updateChannel chan<- interface{}, errorChannel chan<- error, configuration interface{}, watchKey string) {
+	// some watch keys may already have the "/", add it for those that don't
+	if !strings.Contains(watchKey, "/") {
+		watchKey = "/" + watchKey
+	}
+
+	decoder := client.newConsulDecoder()
+	decoder.Consul = client.consulConfig
+	decoder.Target = configuration
+	decoder.Prefix = client.configBasePath + watchKey
+	decoder.ErrCh = errorChannel
+	decoder.UpdateCh = updateChannel
+
+	defer decoder.Close()
+
+	go decoder.Run()
+}
+
+// Simply checks if Consul is up and running at the configured URL
+func (client *consulClient) IsRegistryRunning() bool {
+
+	resp, err := http.Get(client.consulUrl + consulStatusPath)
+	if err != nil {
+		return false
+	}
+
+	if resp.StatusCode >= http.StatusOK && resp.StatusCode < http.StatusMultipleChoices {
+		return true
+	}
+
+	return false
+}
+
+// Checks if a configuration value exists in Consul
+func (client *consulClient) ConfigurationValueExists(name string) (bool, error) {
+	keyPair, _, err := client.consulClient.KV().Get(client.fullPath(name), nil)
+	if err != nil {
+		return false, fmt.Errorf("unable to check existence of %s in Consul: %v", client.fullPath(name), err)
+	}
+	return keyPair != nil, nil
+}
+
+// Gets a specific configuration value from Consul
+func (client *consulClient) GetConfigurationValue(name string) ([]byte, error) {
+	keyPair, _, err := client.consulClient.KV().Get(client.fullPath(name), nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get value for %s from Consul: %v", client.fullPath(name), err)
+	}
+
+	if keyPair == nil {
+		return nil, nil
+	}
+
+	return keyPair.Value, nil
+}
+
+// Puts a specific configuration value into Consul
+func (client *consulClient) PutConfigurationValue(name string, value []byte) error {
+	keyPair := &consulapi.KVPair{
+		Key:   client.fullPath(name),
+		Value: value,
+	}
+
+	_, err := client.consulClient.KV().Put(keyPair, nil)
+	if err != nil {
+		return fmt.Errorf("unable to put value for %s into Consul: %v", client.fullPath(name), err)
+	}
+	return nil
+}
+
+// Gets the service endpoint information for the target ID from Consul
+func (client *consulClient) GetServiceEndpoint(serviceID string) (registry.ServiceEndpoint, error) {
+	services, err := client.consulClient.Agent().Services()
+	if err != nil {
+		return registry.ServiceEndpoint{}, err
+	}
+
+	endpoint := registry.ServiceEndpoint{}
+	if service, ok := services[serviceID]; ok {
+		endpoint.Port = service.Port
+		endpoint.ServiceId = serviceID
+		endpoint.Host = service.Address
+	}
+
+	return endpoint, nil
+}
+
+// Checks with Consul if the target service is available, i.e. registered and healthy
+func (client *consulClient) IsServiceAvailable(serviceID string) bool {
```

Comment:
While integrating into the device-sdk-go I found this function needs to also check the health status and return error when false returned. Will make that change now.